usaspending_api/data_load/fpds_loader.py

```diff
@@ -0,0 +1,340 @@
+
+            # BLIND INSERTS
+            # First create the various records that are shotgunned out on a per-loaded-transaction basis
+            inserted_recipient_locations = _insert_recipient_locations(cursor, load_objects)
+            for index in range(0, len(inserted_recipient_locations)):
+                load_objects[index]["legal_entity"]["location_id"] = inserted_recipient_locations[index][0]
+
+            inserted_recipients = _insert_recipients(cursor, load_objects)
+            for index in range(0, len(inserted_recipients)):
+                load_objects[index]["transaction_normalized"]["recipient_id"] = inserted_recipients[index][0]
+                load_objects[index]["award"]["recipient_id"] = inserted_recipients[index][0]
+
+            inserted_place_of_performance = _insert_place_of_performance(cursor, load_objects)
+            for index in range(0, len(inserted_place_of_performance)):
+                load_objects[index]["transaction_normalized"][
+                    "place_of_performance_id"
+                ] = inserted_place_of_performance[index][0]
+                load_objects[index]["award"]["place_of_performance_id"] = inserted_place_of_performance[index][0]
+
+            # Handle transaction-to-award relationship for each transaction to be loaded
+            for load_object in load_objects:
+
+                # AWARD GET OR CREATE
+                award_id = _lookup_award_by_transaction(cursor, load_object)
+                if not award_id:
+                    # If there is no award, we need to create one
+                    award_id = _insert_award_by_transaction(cursor, load_object)
+
+                load_object["transaction_normalized"]["award_id"] = award_id
+                ids_of_awards_created_or_updated.add(award_id)
+
+                # TRANSACTION UPSERT
+                transaction_id = _lookup_existing_transaction(cursor, load_object)
+                if transaction_id:
+                    # Inject the Primary Key of transaction_normalized+transaction_fpds that was found, so that the
+                    # following updates can find it to update
+                    load_object["transaction_fpds"]["transaction_id"] = transaction_id
+                    _update_fpds_transaction(cursor, load_object, transaction_id)
+                else:
+                    # If there is no transaction we create a new one.
+                    transaction_id = _insert_fpds_transaction(cursor, load_object)
+
+                load_object["transaction_fpds"]["transaction_id"] = transaction_id
+                load_object["award"]["latest_transaction_id"] = transaction_id
+
+                # No matter what, we need to go back and update the award's latest transaction to the award we just made
+                _update_award_latest_transaction(cursor, award_id, transaction_id)
+
+    return list(ids_of_awards_created_or_updated)
+
+
+def _insert_recipient_locations(cursor, load_objects):
```

Comment:
What are the benefits of marking all of these functions as weak-private?

They seem rather clean and clear. No obtuse dependencies or unintended consequences but invoking these functions. 

Additionally, these "load" functions should be stored separately from the "FPDS" code since nothing ties them specifically to FPDS data. They can be imported and used as-is for FABS and perhaps other loaders as well

Reply:
I wanted to mark them as not the part of the loader implementers should be using or entering the file with. Moving them straight out to another file makes plenty of sense though...

Reply:
@tony-sappe I ended up following that trend of "weak-private" as other functions were factored-out.
It's was for encapsulation. To suggest that only code within the module should be calling those functions, rather than someone seeing something like `_insert_recipient_location` as a generally-available function, and start using it for other purposes.

I think this generalized rule is a good one:
- if it has a leading underscore, only call it if you're in the same scope
    - if in a class, if you're another method in that class 
    - if in a module, if you'r another function in that module
- if not, feel free to reuse it in your code in other modules

Reply:
What is the danger of one of those functions being imported and used outside of the FPDS loader function?

Private methods should be to deter misusing a function or to prevent unintended consequences if the behavior is opaque and isn't well implemented.

If they are moved to another file as also suggested (which does make a lot of sense) they cannot be weak-private and adhere to the generalized rule you are proposing since the functions would be imported.

Reply:
marking as resolved since these were moved and re-public-ized

Reply:
@tony-sappe Just to answer the question above
> What is the danger of one of those functions being imported and used outside of the FPDS loader function?

I think the more _intentionally_ reusable a function is made, the less danger this is to have it imported/reused. You had a good argument that there was some utility in these being also used for other loaders. Part of moving them out was getting rid of the side-effects that these functions were doing when "private" -- namely, they were changing the state of the `load_object` dictionary alongside executing the SQL. 

Having a function private also conveys that the author does not want anyone to **re**use them outside their intended scope (class or module), and therefore reserves the right to change the behavior of the functions and add/alter their side-effects, so long as they still serve the author's intended purpose in their class or module.

Basically lessening the scope of use allows for quicker and safer refactoring/evolving of functions.